packages/react-reconciler/src/ReactFiberNewContext.old.js

```diff
@@ -311,6 +329,101 @@ export function propagateContextChange<T>(
   }
 }
 
+export function lazilyPropagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  propagateParentContextChanges(current, workInProgress, renderLanes, false);
+}
+
+export function propagateParentContextChangesToDeferredTree(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  propagateParentContextChanges(current, workInProgress, renderLanes, true);
+}
+
+function propagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+  // TODO: This argument is currently unused. I think there's a way to optimize
+  // for the many providers case, where if the first propagation finds a match,
+  // the second one can avoid scanning down that same path. The trouble is that
+  // there could be a nested bailout below that.
+  forcePropagateEntireTree: boolean,
+) {
+  if (!enableLazyContextPropagation) {
+    return false;
+  }
+
+  let parent = workInProgress;
+  while (parent !== null && (parent.flags & DidPropagateContext) === NoFlags) {
```

Comment:
don't bail out on DidPropagateContext and let the parent loop ride to the root every time.
Then track already propagated providers using a Set or restructure entirely and maintain a "hasChanges" provider set that is added and deleted from on provider push / pop